crates/hir_ty/src/inference.rs

```diff
@@ -224,7 +224,7 @@ impl<'a> InferBody<'a> {
                 self.mut_current_scope().insert(*value, ty_scheme);
                 ty
             }
-            hir::Stmt::ExprStmt {
+            hir::Stmt::Expr {
                 expr,
                 has_semicolon: _,
             } => self.infer_expr(*expr),
```

Comment:
**Optimization Issue**: Recursive inference of expressions with 'infer_expr' can lead to performance issues in cases of deeply nested expressions, similar to the previous issues in expression handling in other parts of the code. <br> **Fix**: Implement an iterative version of 'infer_expr' to handle expression trees iteratively. This would mitigate the risks associated with deep recursion and potentially enhance the performance by reducing call stack usage. <br> **Code Suggestion**: 
 ```
 - self.infer_expr(*expr);
 + self.iterative_infer_expr(*expr);
 ```